### ISSUE TITLE
getting: "error: initializer element is not a compile-time constant" Even when struct is inside function

### ISSUE THREAD

**Issue by aalmkainzi** [llvm/llvm-project#87867]
```C
int foo(int *a, int b)
{
    return 0;
}

int main()
{
    int x;
    (struct {
          typeof(foo(&(struct { int t; }){.t = x}.t, 0)) t;
        }){0};
}
```

the above code compiles with gcc and tcc. but fails with clang.

The error I get is: 
```
err_test.c:10:48: error: initializer element is not a compile-time constant
          typeof(foo(&(struct { int t; }){.t = x}.t, 0)) t;
                                               ^
1 error generated.
```

This error seems to be for global or static variables, but for some reason is triggering even though the variable is defined inside a function and inside a `typeof`

**Comment by shafik** [comment]
`x` does not look like a constant expression in C, I think gcc is accepting that as an implementation defined constant expression.

However if we change `x` to be `const` or `constexpr` and initialize it then both accept it: https://godbolt.org/z/jT6Yzoje3

This makes a lot more sense, although the `const` case feels like an extension.

CC @AaronBallman @cor3ntin @Fznamznon

**Comment by aalmkainzi** [comment]
> `x` does not look like a constant expression in C, I think gcc is accepting that as an implementation defined constant expression.
> 
> However if we change `x` to be `const` or `constexpr` and initialize it then both accept it: https://godbolt.org/z/jT6Yzoje3
> 
> This makes a lot more sense, although the `const` case feels like an extension.
> 
> CC @AaronBallman @cor3ntin @Fznamznon

Why would x be required to be constant in this context?

**Comment by AaronBallman** [comment]
It's not required to be a constant in this case. The compound literal is at block scope and so the initializer does not need to be constant (C23 6.5.3.6p4).

The issue is that we're checking for file scope at https://github.com/llvm/llvm-project/blob/f6357bb4283e72d1248b9c7eb67d98bea71d5f50/clang/lib/Sema/SemaExpr.cpp#L7863 and because the context is "record scope" we assume it must be file scope.

**Comment by aalmkainzi** [comment]
shouldn't also work for file scope? since the compound literal is inside a `typeof`

**Comment by AaronBallman** [comment]
> shouldn't also work for file scope? since the compound literal is inside a `typeof`

There are no special allowances for initialization that occurs in an unevaluated context, so being inside of `typeof` (or `sizeof`, etc) makes no difference.
